### PR TITLE
implement thread-spawn

### PR DESCRIPTION
**Pull request by oligamiq**
I have implemented a multi-thread accessible file system, which was a prerequisite for wasi-threads.

I haven’t created any tests as I’m not quite sure how to do that.

Features:
Usable across multiple workers
Create a WASIFarm and obtain a WASIFarmRef using .get_ref().
WASIFarmRef can be sent via postMessage.
Be cautious, as classes will also be sent as objects via postMessage.
Multiple WASIFarm can be used to create WASIFarmAnimal.
For Stdio, the WASIRef at the end takes precedence.
Files passed to WASIRef will be treated as if they exist.
WASIFarmAnimal can be accessed using .wasiImport.
Each WASIFarmAnimal can only have one .wasiImport.
When creating a WASIFarmAnimal, you can pass override_fd_maps to change the files accessible by each WASIFarmRef. This allows passing file descriptors to child processes.
For usage examples, please refer to the wasi_threads example.
I ported the example rustc and it works fine!
Issues:
Uses SharedArrayBuffer
Since the communication portion is separated, by understanding the polyfill found at https://github.com/WebReflection/sabayon, it should be possible to implement this in a non-SharedArrayBuffer environment.

Commands used across multiple files cannot yet be used across multiple Farms
I will address this eventually.

File pointers are not process-dependent
This could be achieved by using pread to implement read, but I’m not sure if it's correct to make file pointers process-dependent in the first place.

Parts that haven’t been tested may be unreliable
If 100% is the amount of code I can manage, then this project has reached 120%, so I may not be able to manage everything.

It should be correct, but atomics.waitAsync gives tsc error.

Motivation is running low

### CODE REVIEW

**Comment by oligamiq** [comment]
rustc on LLVM confirmed to work properly

**Comment by oligamiq** [comment]
I try running wasm-ld using
https://github.com/YoWASP/clang

**Comment by bjorn3** [comment]
https://github.com/bjorn3/rust/pull/8 links lld directly into the rustc.wasm binary such that it doesn't need to run an external linker executable.

**Comment by oligamiq** [comment]
Have you reached the level of availability yet?

**Comment by bjorn3** [comment]
What do you mean?

**Comment by oligamiq** [comment]
Does it work?

**Comment by bjorn3** [comment]
It works in wasmtime at least:

```shell
$ echo 'fn main() { println!("Hello World!"); }' | wasmtime run -Sthreads=y -Spreview2=n --dir tmp::/ --dir dist --env RUST_MIN_STACK=16777216 dist/bin/rustc.wasm - --sysroot dist --target wasm32-wasip1
Linking using LC_ALL="C" VSLANG="1033" "rust-lld" "-flavor" "wasm" [...]
$ wasmtime run tmp/rust_out.wasm
Hello World!
```

**Comment by oligamiq** [comment]
ok, I'll try it after it's compiled on my server

**Comment by oligamiq** [comment]
Report on the result of using wasm-ld
Clang exits without any output. It might be caused by https://github.com/bjorn3/browser_wasi_shim/issues/80.
When passing a blob pattern to wasm-ld, it directly passes it to `path_open`, which causes an issue (should this be addressed?).
When passing proper arguments, it fails midway with a "move but wasm throw error: memory access out of bounds" error.

I should test wasm-ld with wasmtime, but first, I plan to try https://github.com/bjorn3/rust/pull/8.

**Comment by bjorn3** [comment]
> When passing a blob pattern to wasm-ld, it directly passes it to path_open, which causes an issue (should this be addressed?).

On Unix glob patterns are normally handled by the shell, not by the program itself.

**Comment by oligamiq** [comment]
Using https://github.com/bjorn3/rust/pull/8, I successfully compiled and executed it directly on the web!

**Comment by bjorn3** [comment]
Great work! Just as a heads up: It will almost certainly take me at least a week before I have time to review this.

**Comment by oligamiq** [comment]
I've written quite a lot of comments, but since there are many files and the structure is quite complex, feel free to ask if you have any concerns!

**Comment by oligamiq** [comment]
I don't understand poll_oneoff at all, but since I'm implementing a file system that can be accessed by multiple workers, it's possible that I can implement poll_oneoff as well.

**Comment by bjorn3** [comment]
> I don't understand poll_oneoff at all

It's similar to epoll on Linux, kqueue on BSD. Basically it allows you to ask a bunch of fd's at once if they have any data ready to ready from or can be written too. This is what async rust internally uses on wasi.

File: src/wasi_farm/shared_array_buffer/worker_background/worker_blob.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
...
...("wasi done but worker exists, terminate "+e)),e++;this.start_worker.terminate(),this.start_worker=void 0,console.log("start worker done so terminate")}},this.start_worker.postMessage({...this.override_object,...h,worker_background_ref:this.ref()})}}let i=Atomics.exchange(e,1,0);if(1!==i)throw Error("Lock is already set");let a=Atomics.notify(e,1,1);if(1!==a){if(0===a){console.warn("notify failed, waiter is late");continue}throw Error("notify failed: "+a)}}catch(e){console.error(e)}}constructor(e,r,t,o){_define_property(this,"override_object",void 0),_define_property(this,"allocator",void 0),_define_property(this,"lock",void 0),_define_property(this,"signature_input",void 0),_define_property(this,"workers",[void 0]),_define_property(this,"start_worker",void 0),_define_property(this,"listen_holder",void 0),this.override_object=e,this.lock=r??new SharedArrayBuffer(20),this.allocator=t??new AllocatorUseArrayBuffer(new SharedArrayBuffer(10240)),this.signature_input=o??new SharedArrayBuffer(24),this.listen_holder=this.listen()}};globalThis.onmessage=e=>{let{override_object:r,worker_background_ref_object:t}=e.data;worker_background=WorkerBackground.init_self(r,t),postMessage("ready")};';
```
Comment (bjorn3) [inline src/wasi_farm/shared_array_buffer/worker_background/worker_blob.ts:3]: Is this a compiled version of src/wasi_farm/shared_array_buffer/worker_background/worker.ts?

File: test/run-testsuite.sh
Comment (bjorn3) [inline test/run-testsuite.sh:1]: I think you accidentally changed the file permissions to remove the executable right.

File: .vscode/settings.json
Comment (bjorn3) [inline .vscode/settings.json:1]: Would you mind removing this file?

File: .gitmodules [hunk 2]
```diff
@@ -5,3 +5,7 @@
 	path = test/wasi-testsuite
 	url = https://github.com/WebAssembly/wasi-testsuite
 	branch = prod/testsuite-base
+[submodule "examples/wasi_multi_threads_rustc/rust_wasm"]
+	path = examples/wasi_multi_threads_rustc/rust_wasm
+	url = https://github.com/oligamiq/rust_wasm
+  sharrow = true
```
Comment (bjorn3) [inline .gitmodules:11]:
```suggestion
 	shallow = true
```

**Review by bjorn3 — COMMENTED** [review]
This is a lot more code than I expected. I've left a couple of small comments, but I'm no way near done reviewing this. It will likely take at least a couple of weeks before I've fully reviewed this. Would it be possible to have the thread support in a separate package? If so if it works locally for me, I think it could be merged mostly as is with a disclaimer that it is less production ready than the rest of browser_wasi_shim.

**Comment by oligamiq** [comment]
We are dependent on classes such as Fd, File, and Directory. Since I have only modified src/wasi.ts and have isolated everything in the src/wasi_farm directory, I believe it is quite feasible to create a separate npm package that has @bjorn3/browser_wasi_shim as a dependency.

File: src/wasi_farm/shared_array_buffer/worker_background/worker_blob.ts (hunk 1 above)
Comment (oligamiq) [reply src/wasi_farm/shared_array_buffer/worker_background/worker_blob.ts:3]: Yes

**Comment by oligamiq** [comment]
Regarding the wasm-ification of cargo:
For the time being, I’ve managed to run `cargo run`, so I’ll pause the work for now. I’ve disabled all git-related functionalities, so I can’t use any libraries. Due to the heavy use of threads, the process has become extremely slow. Therefore, I plan to wait for the implementation of thread pools, `fd_filestat_set_times`, and further maturity of gitoxide, as well as WASI Preview 2 reaching Tier 2.

I won’t be submitting a pull request because the current state of cargo is too problematic. However, in the process, I did discover a bug related to this pull request. Additionally, I implemented a special function to run wasm within virtual files, and I’d like to add that as well. Would that be acceptable?

**Comment by bjorn3** [comment]
> Additionally, I implemented a special function to run wasm within virtual files, and I’d like to add that as well. Would that be acceptable?

If you add it outside of the `wasi_snapshot_preview1` module import, it should be possible for the users of browser_wasi_shim to add it without direct browser_wasi_shim support, right? I'm fine with adding it to one of the examples, but I'm hesitant to add extensions to wasi to the core of browser_wasi_shim.

**Comment by oligamiq** [comment]
Indeed. I'll submit the bug fix only.

**Comment by oligamiq** [comment]
> This is a lot more code than I expected. I've left a couple of small comments, but I'm no way near done reviewing this. It will likely take at least a couple of weeks before I've fully reviewed this. Would it be possible to have the thread support in a separate package? If so if it works locally for me, I think it could be merged mostly as is with a disclaimer that it is less production ready than the rest of browser_wasi_shim.

What shall we do?

**Comment by bjorn3** [comment]
> What shall we do?

Please split the thread support into a new package that depends on the main browser_wasi_shim package.

**Comment by oligamiq** [comment]
Would you prefer to go with a monorepo, or should I set up a separate repository?

**Comment by bjorn3** [comment]
Having it in this repo is fine.

**Comment by oligamiq** [comment]
Ok, I try monorepo.
However, it will take time because I don't know the publishing system of this repository.

**Comment by oligamiq** [comment]
By the way, do you have a good package name?

**Comment by bjorn3** [comment]
Maybe browser_wasi_shim-threads?

**Comment by oligamiq** [comment]
I splited, so check please.

File: .gitmodules [hunk 3]
```diff
@@ -5,3 +5,7 @@
 	path = test/wasi-testsuite
 	url = https://github.com/WebAssembly/wasi-testsuite
 	branch = prod/testsuite-base
+[submodule "threads/examples/wasi_multi_threads_rustc/rust_wasm"]
+	path = threads/examples/wasi_multi_threads_rustc/rust_wasm
+	url = https://github.com/oligamiq/rust_wasm
+	sharrow = true
```
Comment (bjorn3) [inline .gitmodules:11]:
```suggestion
	shallow = true
```

File: threads/README.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# A pure javascript shim for WASI Preview 1 threads
+
```
Comment (bjorn3) [inline threads/README.md:2]:
```suggestion

> [!WARNING]
> The code in this directory is less production ready than the main browser_wasi_shim code.

```

File: threads/README.md [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+This project is implement threads on browser_wasi_shim
```
Comment (bjorn3) [inline threads/README.md:3]:
```suggestion
This project implement threads on browser_wasi_shim
```
nit

**Review by bjorn3 — APPROVED** [review]
Couple minor changes and I will merge it.

File: .gitmodules (hunk 3 above)
Comment (oligamiq) [reply .gitmodules:11]: oh...

File: threads/README.md (hunk 4 above)
Comment (oligamiq) [reply threads/README.md:2]: thanks!

Comment (oligamiq) [reply threads/README.md:2]: I'm sorry. I forgot to mention that cross-origin isolation is required. Could you please add that?

**Comment by bjorn3** [comment]
(did you forget to push?)

**Comment by oligamiq** [comment]
Oh, I see. I couldn’t pull because I resolved it without committing. My apologies and thank you so much!

**Comment by bjorn3** [comment]
Thanks a lot for working on this!

**Comment by oligamiq** [comment]
Oh, my apologies, I just finished making the corrections, so it was a bit too early to merge.

**Comment by bjorn3** [comment]
My bad. I fixed the typo you hadn't resolved yet when I merged this PR myself, but for any other changes please open another PR.

**Comment by oligamiq** [comment]
ok.

**Pull request merged**